stan/math/rev/core/vari.hpp

```diff
@@ -198,41 +183,48 @@ using vari = vari_value<double>;
  *
  */
 template <typename T>
-class vari_value<T, std::enable_if_t<is_eigen_dense_base<T>::value>>
-    : public vari_base {
+class vari_value<T, require_eigen_dense_base_t<T>> : public vari_base {
  public:
+  static_assert(
+      is_plain_type<T>::value,
+      "The template for this var is an"
+      " expression but a var_value's inner type must be assignable such as"
+      " a double, Eigen::Matrix, or Eigen::Array");
+
   /**
    * `PlainObject` represents a user constructible type such as Matrix or Array
    */
-  using PlainObject = std::decay_t<plain_type_t<T>>;
+  using PlainObject = plain_type_t<T>;
+  using value_type = PlainObject;  // The underlying type for this class
   using eigen_scalar = value_type_t<PlainObject>;  // A floating point type
+  /**
+   * Maps for adj_ and val_
+   */
+  using eigen_map = Eigen::Map<PlainObject, Eigen::Aligned8>;
+  using vari_type = vari_value<T, require_eigen_dense_base_t<T>>;
   eigen_scalar* val_mem_;  // Pointer to memory allocated on the stack for val_
   eigen_scalar* adj_mem_;  // Pointer to memory allocated on the stack for adj_
-  using Scalar = PlainObject;  // The underlying type for this class
-  using value_type = Scalar;   // The underlying type for this class
-  using eigen_map = Eigen::Map<PlainObject>;  // Maps for adj_ and val_
-  const Eigen::Index rows_;                   // Number of rows
-  const Eigen::Index cols_;                   // Number of cols
-  const Eigen::Index size_;                   // Size of rows * cols
   /**
    * Number of rows known at compile time
    */
-  static constexpr Eigen::Index RowsAtCompileTime = Scalar::RowsAtCompileTime;
+  static constexpr Eigen::Index RowsAtCompileTime
+      = PlainObject::RowsAtCompileTime;
   /**
    * Number of columns known at compile time
    */
-  static constexpr Eigen::Index ColsAtCompileTime = Scalar::ColsAtCompileTime;
+  static constexpr Eigen::Index ColsAtCompileTime
+      = PlainObject::ColsAtCompileTime;
```

Comment:
Eigen uses enums to store these. I like our approach with constexprs better, but I think the type should be int.

Reply:
I thought that enums were preferred since the ```static constexpr``` doesn't always operate as purely compile-time (based on [this writeup](https://www.informit.com/articles/article.aspx?p=2832416&seqNum=6)):
> Static constant members are lvalues (see Appendix B). So, if we have a declaration such as
```void foo(int const&);```
and we pass it the result of a metaprogram:
```foo(Pow3<7>::value);```
a compiler must pass the address of ```Pow3<7>::value```, and that forces the compiler to instantiate and allocate the definition for the static member. As a result, the computation is no longer limited to a pure “compile-time” effect.
Enumeration values aren’t lvalues (i.e., they don’t have an address). So, when we pass them by reference, no static memory is used. It’s almost exactly as if you passed the computed value as a literal. The first edition of this book therefore preferred the use of enumerator constants for this kind of applications.
C++11, however, introduced constexpr static data members, and those are not limited to integral types. They do not solve the address issue raised above, but in spite of that shortcoming they are now a common way to produce results of metaprograms.

Reply:
I did not know that and I have to say I don't fully understand it. How can enums be passed around without taking memory?

Anyway I am fine with changing this to an enum if that is prefered.

Reply:
I definitely don't know enough to have preference here, just something I stumbled on recently that I thought was interesting.

Reply:
Reading the section I think the only concern is when the `static constexpr` is used by a function taking a reference? imo I like the static constexpr version and idt the performance difference here would be drastic with either/or